pipeline_dp/budget_accounting.py

```diff
@@ -100,7 +100,182 @@ def compute_budgets(self):
+    """
+    noise_kind: NoiseKind
+    sensitivity: float = 1
+    weight: float = 1
+    _noise: float = None
+
+    @property
+    def noise(self):
+        """Noise value for the mechanism.
+
+        Raises:
+            AssertionError: The noise value is not calculated yet.
+        """
+        if self._noise is None:
+            raise AssertionError("Noise is not calculated yet.")
+        return self._noise
+
+
+class PLDBudgetAccountant:
+    """Manages the privacy budget for privacy loss distributions."""
+
+    def __init__(self,
+                 total_epsilon: float,
+                 total_delta: float,
+                 pld_discretization: float = 1e-4):
+        """Constructs a PLDBudgetAccountant.
+
+        Stores the mechanisms, the given and calculated attributes
+        of the pipeline.
+
+        Args:
+            total_epsilon: epsilon for the entire pipeline
+            total_delta: delta for the entire pipeline
+            pld_discretization: `value_discretization_interval` in PLD library
+
+        Raises:
+            ValueError: Arguments are missing or out of range.
+        """
+
+        validate_epsilon_delta(total_epsilon, total_delta)
+
+        self._total_epsilon = total_epsilon
+        self._total_delta = total_delta
+        self._mechanisms = []
+        self.minimum_noise = None
+        self._pld_discretization = pld_discretization
+
+    def request_budget(self,
+                       noise_kind: NoiseKind,
+                       sensitivity: float = 1,
+                       weight: float = 1) -> MechanismSpec:
+        """Request a budget.
+
+        Constructs a mechanism spec based on the parameters.
+        Adds the mechanism to the pipeline for future calculation.
+
+        Args:
+            noise_kind: The kind of distribution for the mechanism.
+            sensitivity: The sensitivity for the mechanism.
+            weight: The weight for the mechanism.
+
+        Returns:
+            A mechanism spec based on the parameters.
+        """
+        mechanism_spec = MechanismSpec(noise_kind=noise_kind,
+                                       sensitivity=sensitivity,
+                                       weight=weight)
+        self._mechanisms.append(mechanism_spec)
+        return mechanism_spec
+
+    def compute_budgets(self):
+        """Computes the budget for the pipeline.
+
+        Composes the mechanisms and adjusts the amount of
+        noise based on given epsilon. Sets the noise for the
+        entire pipeline.
+        """
+        if not self._mechanisms:
+            return
+        minimum_noise = self._compute_and_adjust()
+        self.minimum_noise = minimum_noise
+        for mechanism in self._mechanisms:
+            mechanism._noise = minimum_noise / mechanism.weight
+
+    def _compute_and_adjust(self) -> float:
+        """ Adjust the noise value to the minimum.
+
+        Repeat the PLD composition until the noise value gives a
```

Comment:
Nit: Repeat the PLD composition ->  Uses binary search to ...   (it's more precise)